client-cert: certificate chain in the same order as it appears in TLS (#2258)
httpwg/http-extensions: draft-ietf-httpbis-client-cert-field.md

Client-Cert-Chain:
: Conveys the certificate chain used for validation of the end-entity

> certificate provided by the client in the TLS handshake with the reverse proxy from the proxy to the origin server.

Reviewer: This sentence has gotten a little hard to follow.

There are two things here that might need to be separated to make this more readable:

1. This is the chain (less the EE cert) that the client presents in TLS to the TTRP.
2. The TTRP sends this information to the origin server.

Reviewer 2: I know. I know. Saying it is "a little hard to follow" is being too kind. It's pretty terrible. And I seem to make it worse each time I touch it. I'll take another pass over it though with that separation in mind. Maybe leaning more on the text above to make the second point and leaving it out of the Client-Cert/Client-Cert-Chain items.

## Client-Cert-Chain HTTP Header Field {#chain-header}

In the context of a TLS terminating reverse proxy deployment, the proxy
MAY make the certificate chain for validation of the end-entity certificate
available to the backend application with the Client-Cert-Chain HTTP header
field. This field contains certificates for validation of the end-entity
certificate provided by the client in the TLS handshake. These certificates might or

> might not have been provided by the client during the TLS handshake, however,

Reviewer: Can you explain your reasoning behind this "might or might not" clause?

Author: That text was from 0a7210f7d644dea399f0fc4d1b5119acb436d1db when @MikeBishop added the Client-Cert-Chain header and I was trying not to step on his work too much. And also to allow for the case that the proxy assembles the chain and conveys it.

Reviewer: This seems problematic as it invites a query about why that might (or might not) be the case.

I would prefer to say that the proxy can add certificates to this field.  It provides the certificates in the order in which they appear in the TLS handshake, unless it has cause to provide other information.

Author: What about even less invitation of questions by making that last sentence this? 

"The order of the certificates is the same as the ordering in TLS (such as described in {{Section 4.4.2 of RFC8446}})."

Reviewer: That would work for me, but I got the impression that you are aware of cases where that is not the case.  I don't think that there is any harm in saying "..., unless the server and proxy agree otherwise." or similar.

Reviewer 3: Honestly don't remember what my logic was in writing that; it appears to have been from my first pass at this text and not touched in any of the feedback.

I *think* that was based on TLS 1.3 saying that "...a certificate that specifies a trust anchor MAY be omitted from the chain, provided that supported peers are known to possess any omitted certificates."  If the proxy possesses certificates the origin might not, the client might omit them from the handshake, but the proxy could append them for the origin's information.

I don't think there's harm in allowing for private agreement here, but I don't feel strongly either way.

Author: As I remember things, there was some disagreement when the initial text for `Cient-Cert-Chain` was added/proposed about whether the proxy would construct and validate the chain and convey the constructed chain to the origin or whether the proxy would pass through what the client sent as-is and defer validation and construction to the origin. Some of this discussion is captured in PR #1874. AFAIK most of the time the chain would look the same. But I've tried to adjust the text such that either would be okay. The sentence above about the order being the same was intended to mean only that ordering is the same but not that they are the same.

Reviewer 2: [5fb237a](https://github.com/httpwg/http-extensions/pull/2258/commits/5fb237a7dbb030ab323710a1d254d442cd41f35d) is another pass at this that removes the "might or might not" and tries to articulate ordering.

## Client-Cert-Chain HTTP Header Field {#chain-header}

In the context of a TLS terminating reverse proxy deployment, the proxy
MAY make the certificate chain
available to the backend application with the Client-Cert-Chain HTTP header
field.

Client-Cert-Chain is a List Structured Header {{!RFC8941}}.  Each item in the
list MUST be a Byte Sequence ({{Section 3.3.5 of RFC8941}}) encoded as described

> in {{encoding}}. The order is the same as the ordering in TLS (such as described in {{Section 4.4.2 of RFC8446}})

Reviewer: ```suggestion
in {{encoding}}. The order is the same as the ordering in TLS (such as described in {{Section 4.4.2 of RFC8446}}).
```

certificate information, this draft opted to pass the full encoded certificate
as the value of the `Client-Cert` field.

The validation of the client certificate and chain of the mutually-authenticated
TLS connection is typically performed by the TTRP during the handshake.  With the
responsibility of certificate validation falling on the TTRP, the
end-entity certificate is oftentimes sufficient for the needs of the origin server.
The separate `Client-Cert-Chain` field can convey the certificate chain for

> deployments that require such information in order for the origin server to augment
> the chain validation or assume it entirely.

Reviewer: ```suggestion
origin server deployments that require this additional information.
```

Maybe simpler is better.

Reviewer 2: Maybe so.
